core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java

```diff
@@ -194,49 +199,84 @@ private static ServiceRequestContext serviceRequestContext() {
      *         completed with the cause of the failure.
      */
-    public boolean init(EndpointGroup endpointGroup) {
+    public CompletableFuture<Boolean> init(EndpointGroup endpointGroup) {
         assert endpoint == null : endpoint;
         assert !initialized;
         initialized = true;
 
-        Throwable cause = null;
         try {
             if (endpointGroup instanceof Endpoint) {
-                this.endpointGroup = null;
-                updateEndpoint((Endpoint) endpointGroup);
-                runThreadLocalContextCustomizers();
+                return initEndpoint((Endpoint) endpointGroup);
             } else {
-                this.endpointGroup = endpointGroup;
-                // Note: thread-local customizer must be run before EndpointSelector.select()
-                //       so that the customizer can inject the attributes which may be required
-                //       by the EndpointSelector.
-                runThreadLocalContextCustomizers();
-                updateEndpoint(endpointGroup.select(this));
+                return initEndpointGroup(endpointGroup);
             }
         } catch (Throwable t) {
-            cause = t;
+            acquireEventLoop(endpointGroup);
+            failEarly(t);
+            return UnmodifiableFuture.completedFuture(false);
         }
+    }
 
-        if (eventLoop == null) {
-            final ReleasableHolder<EventLoop> releasableEventLoop =
-                    options().factory().acquireEventLoop(sessionProtocol(), endpointGroup, endpoint);
-            eventLoop = releasableEventLoop.get();
-            log.whenComplete().thenAccept(unused -> releasableEventLoop.release());
+    private UnmodifiableFuture<Boolean> initEndpoint(Endpoint endpoint) {
+        this.endpointGroup = null;
+        updateEndpoint(endpoint);
+        runThreadLocalContextCustomizers();
+        acquireEventLoop(endpoint);
+        return UnmodifiableFuture.completedFuture(true);
+    }
+
+    private CompletableFuture<Boolean> initEndpointGroup(EndpointGroup endpointGroup) {
+        this.endpointGroup = endpointGroup;
+        // Note: thread-local customizer must be run before EndpointSelector.select()
+        //       so that the customizer can inject the attributes which may be required
+        //       by the EndpointSelector.
+        runThreadLocalContextCustomizers();
+        final Endpoint endpoint = endpointGroup.selectNow(this);
+        if (endpoint != null) {
+            updateEndpoint(endpoint);
+            acquireEventLoop(endpointGroup);
+            return UnmodifiableFuture.completedFuture(true);
         }
 
-        if (cause != null) {
-            failEarly(cause);
-            return false;
-        }
+        // Use an arbitrary event loop for asynchronous Endpoint selection.
+        final EventLoop temporaryEventLoop = options().factory().eventLoopSupplier().get();
```

Comment:
Can't we do `acquireEventLoop(endpointGroup);` first and use the eventLoop?

Reply:
We can, but why?

Reply:
Forget about this. 😅